File: NewsDaily2/app/src/main/java/com/andruid/magic/newsdaily/headlines/NewsDataSource.java
package com.andruid.magic.newsdaily.headlines;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;

import com.andruid.magic.newsloader.model.News;
import com.andruid.magic.newsloader.repo.NewsRepository;

import java.util.List;

import timber.log.Timber;

import static android.nfc.tech.MifareUltralight.PAGE_SIZE;

public class NewsDataSource extends PageKeyedDataSource<Integer, News> {
    private static final int FIRST_PAGE = 1;
    private String country, category;

    NewsDataSource(String country, String category) {
        this.category = category;
        this.country = country;
        Timber.d("data source created");
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Integer> params, @NonNull final LoadInitialCallback<Integer, News> callback) {
        Timber.d("load initial");
        NewsRepository.getInstance().loadHeadlines(country, category, FIRST_PAGE, PAGE_SIZE, new NewsRepository.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                callback.onResult(newsList, null, FIRST_PAGE + 1);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }

    @Override
    public void loadBefore(@NonNull final LoadParams<Integer> params, @NonNull final LoadCallback<Integer, News> callback) {
        NewsRepository.getInstance().loadHeadlines(country, category, params.key, PAGE_SIZE, new NewsRepository.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                Integer adjacentKey = (params.key > FIRST_PAGE) ? params.key - 1 : null;
                if(newsList!=null)
                    callback.onResult(newsList, adjacentKey);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }

    @Override
    public void loadAfter(@NonNull final LoadParams<Integer> params, @NonNull final LoadCallback<Integer, News> callback) {
        Timber.d("load page %d", params.key);
        NewsRepository.getInstance().loadHeadlines(country, category, params.key, PAGE_SIZE, new NewsRepository.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                Integer key = hasMore ? params.key + 1 : null;
                callback.onResult(newsList, key);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/headlines/NewsDataSource.kt
package com.andruid.magic.newsdaily.headlines

import androidx.paging.PageKeyedDataSource
import com.andruid.magic.newsloader.api.NewsRepository.Companion.getInstance
import com.andruid.magic.newsloader.api.NewsRepository.NewsLoadedListener
import com.andruid.magic.newsloader.data.Constants
import com.andruid.magic.newsloader.model.News
import timber.log.Timber

class NewsDataSource(val country : String, val category : String) : PageKeyedDataSource<Int, News>() {

    override fun loadInitial(params: LoadInitialParams<Int>, callback: LoadInitialCallback<Int, News>) {
        getInstance().loadHeadlines(country, category, Constants.FIRST_PAGE, Constants.PAGE_SIZE, object : NewsLoadedListener {
            override fun onSuccess(newsList: List<News>, hasMore: Boolean) {
                callback.onResult(newsList, null, Constants.FIRST_PAGE + 1)
            }

            override fun onFailure(t: Throwable?) {
                t!!.printStackTrace()
            }
        })
    }

    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, News>) {
        Timber.d("load page %d", params.key)
        getInstance().loadHeadlines(country, category, params.key, Constants.PAGE_SIZE, object : NewsLoadedListener {
            override fun onSuccess(newsList: List<News>, hasMore: Boolean) {
                val key = if (hasMore) params.key + 1 else null
                callback.onResult(newsList, key)
            }

            override fun onFailure(t: Throwable?) {
                t!!.printStackTrace()
            }
        })
    }

    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, News>) {
        getInstance().loadHeadlines(country, category, params.key, Constants.PAGE_SIZE, object : NewsLoadedListener {
            override fun onSuccess(newsList: List<News>, hasMore: Boolean) {
                val adjacentKey = if (params.key > Constants.FIRST_PAGE) params.key - 1 else null
                callback.onResult(newsList, adjacentKey)
            }

            override fun onFailure(t: Throwable?) {
                t!!.printStackTrace()
            }
        })
    }
}